pytorch_lightning/overrides/data_parallel.py

```diff
@@ -45,178 +40,132 @@ def _find_tensors(obj):  # pragma: no-cover
-        This class is used in combination with :class:`~torch.nn.parallel.DistributedDataParallel` as
-        shown in the example.
-
-        Example:
-
-            ddp_model = DistributedDataParallel(
-                module=LightningDistributedModule(lightning_module),
-                device_ids=[local_rank],
-                ...
-            )
+        method, either ``training_step``, ``validation_step`` or ``test_step``.
+        If the LightningModule is in none of the states `training`, `testing` or `validation`,
+        the inputs will be redirected to the
+        :meth:`~pytorch_lightning.core.lightning.LightningModule.predict` method.
+        Inheriting classes may also modify the inputs or outputs of forward.
 
         Args:
             pl_module: the model to wrap
-
         """
         super().__init__()
         self.module = pl_module
 
     def forward(self, *inputs, **kwargs):
-
         running_stage = self.module.running_stage
 
         if running_stage == RunningStage.TRAINING:
             output = self.module.training_step(*inputs, **kwargs)
             warn_if_output_is_none(output, "training_step")
-
         elif running_stage == RunningStage.TESTING:
             output = self.module.test_step(*inputs, **kwargs)
             warn_if_output_is_none(output, "test_step")
-
         elif running_stage == RunningStage.EVALUATING:
             output = self.module.validation_step(*inputs, **kwargs)
             warn_if_output_is_none(output, "validation_step")
-
         else:
             output = self.module.predict(*inputs, **kwargs)
 
         return output
 
 
+class LightningParallelModule(_LightningModuleWrapperBase):
+    """
+    Wraps the user's LightningModule and redirects the forward call to the appropriate
+    method, either ``training_step``, ``validation_step``, ``test_step`` or ``predict``.
+    This class is used in combination with :class:`~torch.nn.parallel.DataParallel` as
+    shown in the example. It also takes care of converting Python scalars to Tensors and
+    un-squeezes 0-dimensional Tensors as it is required by :class:`~torch.nn.parallel.DataParallel`.
+
+    Example:
+
+        dp_model = torch.nn.DataParallel(
+            module=LightningParallelModule(lightning_module),
+            device_ids=[3, 4],
+            ...
+        )
+
+    Args:
+        pl_module: the model to wrap
+
+    """
+    def __init__(self, pl_module: LightningModule):
+        super().__init__(pl_module)
+
+    def forward(self, *inputs, **kwargs):
+        output = super().forward(*inputs, **kwargs)
+        output = apply_to_collection(
+            output,
+            dtype=numbers.Number,
+            function=python_scalar_to_tensor,
+            device=self.module.device
+        )
+        output = apply_to_collection(
+            output,
+            dtype=torch.Tensor,
+            function=unsqueeze_scalar_tensor,
+        )
+        return output
```

Comment:
can we maybe combine these functions? then we only need to traverse the collections once